### PR TITLE
js: semgrep js pattern testing

### PR DESCRIPTION
**Pull request by ajbt200128**
This PR does a bunch of things:
- Semgrep.js parsers:
  - Factors out the test harness plumbing into `js/shared/parser.js`
  - Updates parsers to use official pattern / parsing test fixtures, which means more test coverage for jsoo
  - Fixed a few parsers that didn't match their Parse_target2.ml / Parse_pattern2.ml counterparts
- fixed issues with utf-8 in libpcre.js
- jsoo's caml_float_of_string doesn't handle whitespace the same way as Stdlib (issue: https://github.com/ocsigen/js_of_ocaml/issues/1518)
- testing:
  - added `js/tests` which runs the entire Semgrep test suite against the semgrep.js artifacts
    - (with some tests currently disabled for various reasons, see: https://github.com/returntocorp/semgrep/pull/8902/files#diff-dd182dfc008e74db2491c9b572d2f1885943808bf40dd7f2d384a097c6f86a96R26-R39)
- due to how semgrep.js parsers work, had to make jsonnet parser initialization lazy (See: https://github.com/returntocorp/semgrep/pull/8902/files#diff-2ca9fa192a8e7f7613f16ab4969198be90b75bd5fe38ed3f79a1d23fb59f1e59R57-R59)
- Updated Entropy.ml to explicitly use Int64 so that it still works on 32-bit systems (namely: browsers)

(^ @tpetr wrote this -- @ajbt200128 , feel free to update)

test plan:
- checks pass

### CODE REVIEW

**Comment by ajbt200128** [comment]
Prolly want to try and enable these tests in GHA before merging

File: js/engine/src/index.js
Comment (tpetr) [inline js/engine/src/index.js]: shouldnt this be `getLangs()` (capital L)?

File: js/languages/cairo/tests/index.test.js
Comment (tpetr) [inline js/languages/cairo/tests/index.test.js]: can we uncomment this? I just commented it out because I didn't have a good cairo pattern to use in the moment (and then never circled back to fixing). Ditto to the other commented out pattern tests below.

File: js/languages/dockerfile/tests/index.test.js
Comment (tpetr) [inline js/languages/dockerfile/tests/index.test.js]: whats going on here? copypasta?

File: src/optimizing/Analyze_rule.ml
Comment (tpetr) [inline src/optimizing/Analyze_rule.ml:518]: I'm fine with ignoring, but do we know why it fails in jsoo?

File: src/engine/Test_engine.mli
Comment (tpetr) [inline src/engine/Test_engine.mli:8]: is there a tl;dr for what this does? prepends the name of the lang in the summary?

File: src/engine/Entropy.ml
Comment (tpetr) [inline src/engine/Entropy.ml]: we should clean up these printfs before merging

File: js/tests/dune
Comment (tpetr) [inline js/tests/dune:4]: good call, feel free to crank up the verbosity / user friendliness of the jsoo output in general -- eslint will handle minification

File: js/shared/dune
Comment (tpetr) [inline js/shared/dune:13]: what is this?

File: js/shared/Semgrep_js_shared.ml
Comment (tpetr) [inline js/shared/Semgrep_js_shared.ml:64]: where's the `Lang.Json` case? (if we don't actually need it, lets update the comment above). Ditto for `setJustParseWithLang`

File: js/shared/Semgrep_js_shared.ml
Comment (tpetr) [inline js/shared/Semgrep_js_shared.ml:128]: I really struggle with this name -- my mind jumps to "typescript" before "tree-sitter". Can you make the name a little more descriptive?

File: js/libyaml/runtime.js
Comment (tpetr) [inline js/libyaml/runtime.js:46]: we should file a bug against jsoo

File: libs/commons/common2.ml
Comment (tpetr) [inline libs/commons/common2.ml]: I worry about the portability of this. Can we use `Fpath`?

**Review by tpetr — CHANGES_REQUESTED** [review]
Great work, happy to approve once the bigger pieces of feedback are addressed

File: js/engine/src/index.js
Comment (ajbt200128) [reply js/engine/src/index.js]: Yea

File: js/languages/cairo/tests/index.test.js
Comment (ajbt200128) [reply js/languages/cairo/tests/index.test.js]: Up to you, didn't want to mess with it

File: js/languages/dockerfile/tests/index.test.js
Comment (ajbt200128) [reply js/languages/dockerfile/tests/index.test.js]: Yes, feel free to delete the duplicate code

File: src/optimizing/Analyze_rule.ml
Comment (ajbt200128) [reply src/optimizing/Analyze_rule.ml:518]: No, I tried looking for a bit but I just couldn't bring myself to care. The error was "dump(1000)" impossible tag or something similar

File: src/engine/Test_engine.mli
Comment (ajbt200128) [reply src/engine/Test_engine.mli:8]: Yea, that way we can filter out the tests based on languages, in order to skip C++,  julia etc.

Comment (ajbt200128) [reply src/engine/Test_engine.mli:8]: Also easier to read

File: src/engine/Entropy.ml
Comment (ajbt200128) [reply src/engine/Entropy.ml]: Agree

File: js/tests/dune
Comment (ajbt200128) [reply js/tests/dune:4]: Oh true, didn't think about that.

File: js/shared/dune
Comment (ajbt200128) [reply js/shared/dune:13]: semgrep.data is needed so we can call Data.init or smthn in the init functionin Semgrep_js_shared. This is what loads the trigrams for entropy analysis

File: js/shared/Semgrep_js_shared.ml
Comment (ajbt200128) [reply js/shared/Semgrep_js_shared.ml:64]: Yea you're right, no Lang.Json, so comment needs to be updated

File: js/shared/Semgrep_js_shared.ml
Comment (ajbt200128) [reply js/shared/Semgrep_js_shared.ml:128]: I'm ok w/renaming it

File: js/libyaml/runtime.js
Comment (ajbt200128) [reply js/libyaml/runtime.js:46]: Maybe? Idk if the maintainer wouldn't want to break existing jsoo code. I'm ok filing a bug but I don't have any strong feelings

File: libs/commons/common2.ml
Comment (ajbt200128) [reply libs/commons/common2.ml]: Where would we use Fpath? Not sure what's not portable here

Comment (tpetr) [reply libs/commons/common2.ml]: Chatted on Slack, no portability issue here

File: js/shared/Semgrep_js_shared.ml
Comment (tpetr) [reply js/shared/Semgrep_js_shared.ml:128]: I'm kinda struggling with a better name, so let's revisit this in the future

**Review by tpetr — APPROVED** [review]
LGTM

**Comment by tpetr** [comment]
Open issues: (none explicitly blocking AFAIK)
- Semgrep Jsonnet is still unhappy with `Test_jsoo.ml`: ([logs](https://github.com/returntocorp/semgrep/actions/runs/6547440599/job/17779928456?pr=8902)) <-- maybe we just merge anyway
- LLVM build slowness (waiting on next steps for https://github.com/llvm/llvm-project/pull/67715)
- Int overflow bug for greater than 32-bit ints inside metavariable-comparison (similar to the Entropy.ml thing)

File: js/libyaml/runtime.js
Comment (tpetr) [reply js/libyaml/runtime.js:46]: https://github.com/ocsigen/js_of_ocaml/issues/1518, but nbd if they won't fix

**Comment by tpetr** [comment]
Need a codeowner to approve to merge 😞 -- @aryx can you take a look when you have a min?

File: .github/workflows/build-test-javascript.jsonnet
Comment (aryx) [inline .github/workflows/build-test-javascript.jsonnet]: Why not run those tests as part of the 'make test' of the previous step?
make -C js test
seems to go in different subdirs and run 'make test' in it, so maybe he should also go in
the js/tests/ subdir?

File: Makefile
Comment (aryx) [inline Makefile]: This is not needed anymore, the .github/workflow/Makefile now uses directly the _build/... path

Comment (aryx) [reply Makefile]: But can be useful to be consistent with some other build-xxx target.

File: js/engine/Main.ml
Comment (aryx) [inline js/engine/Main.ml]: This is hard to read. Where is defined this setParsePattern on the rhs?

File: js/engine/Main.ml
Comment (aryx) [inline js/engine/Main.ml:8]: please avoid some open if you can. The Js_of_ocaml is ok, the Semgrep_js_shared a bit less.

File: js/engine/Makefile.include
Comment (aryx) [inline js/engine/Makefile.include:15]: what is semgrep-engine.js? It seems to depend only on libyaml.o and libpcre.o so seem to have nothing to do
with semgrep, so why is it called semgrep-engine.js?

File: js/languages/jsonnet/Parser.ml
Comment (aryx) [inline js/languages/jsonnet/Parser.ml]: Please do not prefix function with _ if they are actually used. This is not Python. In OCaml the convention is to prefix a var or func with _ when it's not used but for some reasons we still want to keep in the code, but this prevent ocamlc to issue a warning about unused var or func.

File: js/languages/protobuf/tests/index.test.js
Comment (aryx) [inline js/languages/protobuf/tests/index.test.js]: it's because we didn't define semgrep extensions for protobuf yet I think.

File: js/languages/shared/parser.js
Comment (aryx) [inline js/languages/shared/parser.js:1]: I would have rename shared to SHARED maybe to make it more visible this directory has a different nature
than the other directory in languages/
(or even put it under js/shared/ with the other shared stuff).

File: js/tests/Test_jsoo.ml
Comment (aryx) [inline js/tests/Test_jsoo.ml]: Small comment explaining this constant (and maybe even what is its type).

File: js/tests/Test_jsoo.ml
Comment (aryx) [inline js/tests/Test_jsoo.ml]: please avoid open and use qualified name when possible.

File: js/tests/dune
Comment (aryx) [inline js/tests/dune:4]: maybe a small comment explaining why we need those flags.

File: js/tests/index.test.js
Comment (aryx) [inline js/tests/index.test.js:68]: why we need this special thing for jsonnet? Maybe a comment?

File: js/tests/index.test.js
Comment (aryx) [inline js/tests/index.test.js:101]: maybe a process.exit(0) somewhere.

File: libs/ojsonnet/Parse_jsonnet.mli
Comment (aryx) [inline libs/ojsonnet/Parse_jsonnet.mli:5]: maybe a comment explaining why we need to do that, why we can't call directly Parse_jsonnet_tree_sitter.parse ?

File: src/engine/Test_engine.ml
Comment (aryx) [inline src/engine/Test_engine.ml]: do not use match for boolean; use if then else.

File: src/engine/Test_engine.ml
Comment (aryx) [inline src/engine/Test_engine.ml:307]: why do we need this new code?

File: src/parsing/Parse_target.ml
Comment (aryx) [inline src/parsing/Parse_target.ml]: nice
maybe add a comment like "(* that does not mean we will not use Tree_sitter for parsing JS files
in the engine, because at some point the guard above about just_parse_with_lang_ref will be set to something differnet than undefined_just_parse_with_lang*)"

**Review by aryx — APPROVED** [review]
Nice!

**Comment by aryx** [comment]
I've added you as an admin Tom for this repo so you can force merge if needed.

File: js/tests/Test_jsoo.ml
Comment (ajbt200128) [reply js/tests/Test_jsoo.ml]: Yea would appreciate an explanation of what the array in the tuple represents

**Comment by ajbt200128** [comment]
Double checked, looks gucci

File: Makefile
Comment (tpetr) [reply Makefile]: I'll remove this one-off (added before your ojsonnet makefile tweak) and we can open a separate PR for consistency

File: js/languages/shared/parser.js
Comment (tpetr) [reply js/languages/shared/parser.js:1]: Yeah I'd like to hoist `languages/shared` in general so that people don't think it's a parser. We can do that in a separate PR

File: js/tests/Test_jsoo.ml
Comment (tpetr) [reply js/tests/Test_jsoo.ml]: Will add a comment -- I'm also going to rename the variable to `skipped_tests` for clarity

File: js/tests/dune
Comment (tpetr) [reply js/tests/dune:4]: They were for debugging, but it's possible we don't even need those flags anymore? (any opinions @ajbt200128 ?)

File: libs/ojsonnet/Parse_jsonnet.mli
Comment (tpetr) [reply libs/ojsonnet/Parse_jsonnet.mli:5]: @ajbt200128 can you refresh my memory here? because semgrep eagerly tries to instantiate the jsonnet parser?

File: js/tests/index.test.js
Comment (tpetr) [reply js/tests/index.test.js:101]: You get it for free w/ the method returning, but I can add an explicit one for clarity

File: src/parsing/Parse_target.ml
Comment (tpetr) [reply src/parsing/Parse_target.ml]: Updated comment

File: js/engine/Main.ml
Comment (tpetr) [reply js/engine/Main.ml]: Fixed via getting rid of `open Semgrep_js_shared`

File: js/engine/Makefile.include
Comment (tpetr) [reply js/engine/Makefile.include:15]: `semgrep-engine.js` is JS plumbing for Semgrep's non-tree-sitter WASM dependencies. I agree it's a little confusing -- let's rename (and/or break into `libyaml.js`, `libpcre.js`) in a separate PR.

File: js/languages/protobuf/tests/index.test.js
Comment (tpetr) [reply js/languages/protobuf/tests/index.test.js]: Updated comment to point to the protobuf github issue

File: src/engine/Test_engine.ml
Comment (tpetr) [reply src/engine/Test_engine.ml:307]: that block of code optionally (controlled by `prepend_lang`) prefixes the test name with the lang that it's operating on, so that the test output is easier to read and it's easier for us to skip tests by language

File: libs/ojsonnet/Parse_jsonnet.mli
Comment (ajbt200128) [reply libs/ojsonnet/Parse_jsonnet.mli:5]: Right, semgrep tries to create the tree sitter parser on startup

**Comment by tpetr** [comment]
Merging in spite of the failing lint check because it's a known issue w/ Semgrep Jsonnet (see: https://github.com/returntocorp/semgrep/issues/8792)

**Comment by ihji** [comment]
@tpetr I have concerns about this. It appears that this commit has caused the lint checks for every other PR to fail since it was merged.

**Comment by aryx** [comment]
I've fixed the mistake that was (badly) reported in Test_jsoo.ml in another PR, so we're good.

File: .github/workflows/build-test-javascript.jsonnet
Comment (aryx) [inline .github/workflows/build-test-javascript.jsonnet:138]: don't need this line here; the make test above already run the tests under js/tests now.

File: js/Makefile
Comment (aryx) [inline js/Makefile:17]: maybe better to be consistent and use make test for the tests/ dir too.

File: js/tests/Test_jsoo.ml
Comment (aryx) [inline js/tests/Test_jsoo.ml:52]: that was the location osemgrep was complaining (badly) about. In OCaml == is a physical equality, not structureal equality

File: js/tests/index.test.js
Comment (aryx) [inline js/tests/index.test.js:98]: hmm this finally will never execute then because of the process.exit(0) above.

File: src/optimizing/Analyze_rule.ml
Comment (aryx) [reply src/optimizing/Analyze_rule.ml:518]: Dumper.dump use the Obj module so make sense it fails with jsoo.

File: js/tests/index.test.js
Comment (tpetr) [reply js/tests/index.test.js:98]: Ah yep, was moving too fast. I'm not even sure if we need the logging anymore -- I think that was more of an Austin debugging thing

**Pull request merged**
